File: Java/Platform/Basics/MonitorTest/Program3.java
class Program {

	public static void main(String[] args) throws Exception {
		var acc = new JointAccount();
		acc.deposit(5000);
		System.out.println("Account opened for Jack and Jill.");
		System.out.printf("Initial Balance = %d%n", acc.balance());
		var child = new Thread(() -> {
			System.out.println("Jack is withdrawing 3000...");
			if(acc.withdraw(3000) == false)
				System.out.println("Jack's withdraw operation failed!");
		});
		child.start();
		System.out.println("Jill is withdrawing 4000...");
		if(acc.withdraw(4000) == false)
			System.out.println("Jill's withdraw operation failed!");
		child.join(); //current thread waits for child to exit
		System.out.printf("Final Balance   = %d%n", acc.balance());
	}

	static class JointAccount {
		
		private int balance;

		public int balance() { return balance; }

		public synchronized void deposit(int amount) {
			balance = Worker.doWork(amount, balance, 1);
			//notify the monitor of this object to allow any one
			//thread waiting on it to resume execution
			this.notify();
		}

		public boolean withdraw(int amount) {
			boolean success = false;
			//A thread can enter a synchronized block by locking the built-in
			//monitor of 'this' object. Only one thread can lock the monitor
			//of a given object at a time while other threads must wait for
			//this thread to unlock the monitor by exiting synchronized block
			synchronized(this){
				if(balance >= amount){
					balance = Worker.doWork(amount, balance, -1);
					success = true;
				}
			}
			return success;
		}

		public synchronized boolean withdrawAfterDeposit(int amount) throws InterruptedException {
			//unlock the monitor of 'this' and wait for some other
			//thread to notify this monitor
			this.wait();
			return withdraw(amount);
		}
	}
}

File: dotnet/Platform/Basics/MonitorTest/DemoApp/Program3.cs
using System.Runtime.CompilerServices;

var acc = new JointAccount();
acc.Deposit(5000);
Console.WriteLine("Account opened for Jack and Jill.");
Console.WriteLine("Initial Balance = {0}", acc.Balance);
Thread child = new Thread(() => 
{
    Console.WriteLine("Jack is withdrawing 3000...");
    if(acc.Withdraw(3000) == false)
        Console.WriteLine("Jack's withdraw operation failed!");
});
child.Start();
Console.WriteLine("Jill is withdrawing 4000...");
if(acc.Withdraw(4000) == false)
    Console.WriteLine("Jill's withdraw operation failed!");
child.Join(); //wait for child to exit
Console.WriteLine("Final Balance   = {0}", acc.Balance);


class JointAccount
{
    public int Balance { get; set; }

    public void Deposit(int amount)
    {
        lock(this)
        {
            Balance = Worker.DoWork(amount, Balance, 1);
            //Pulse (signal) the monitor associated with the specified object
            Monitor.Pulse(this);
        }
    }

    public bool Withdraw(int amount)
    {
        bool success = false;
        //Lock the (reentrant) monitor associated with the specified (this)
        //object before proceeding. Only one thread can lock a given monitor
        //at a time while other threads must wait for this thread to unlock
        //that monitor
        Monitor.Enter(this);
        try
        {
            if(Balance >= amount)
            {
                Balance = Worker.DoWork(amount, Balance, -1);
                success = true;
            }
        }
        finally
        {
            //unlock the monitor associated with the specified (this) object
            Monitor.Exit(this);
        }
        return success;
    }

    [MethodImpl(MethodImplOptions.Synchronized)]
    public bool WithdrawAfterDeposit(int amount)
    {
        //Unlock the monitor assocaied with the specified object, 
        //wait for another thread to pulse this monitor, 
        //re-lock the monitor and resume
        Monitor.Wait(this);
        return Withdraw(amount);
    }
}
